sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/joinTypes.scala

```diff
@@ -24,7 +24,7 @@ import org.apache.spark.sql.catalyst.expressions.Attribute
 object JoinType {
   def apply(typ: String): JoinType = typ.toLowerCase(Locale.ROOT).replace("_", "") match {
     case "inner" => Inner
-    case "outer" | "full" | "fullouter" => FullOuter
+    case "outer" | "full" | "fullouter" | "full_outer" => FullOuter
```

Comment:
as @wzhfy points out, given that we remove underscores, this matching case is unnecessary. The same is true for `left_outer`, `right_outer` and `left_semi`.